libs/javalib/worker/src/mill/javalib/worker/JvmWorkerImpl.scala

```diff
@@ -444,623 +63,378 @@ class JvmWorkerImpl(
-          |
-          |    Task.reporter.apply(hashCode)
-          |
-          |`hashCode` should be the hashCode of the Mill module being compiled. Calling
-          |this in a task defined in a module should work.
-          |""".stripMargin
-      )
-
-    os.makeDir.all(ctx.dest)
-
-    val classesDir =
-      if (compileToJar) ctx.dest / "classes.jar"
-      else ctx.dest / "classes"
-
-    if (ctx.log.debugEnabled) {
-      ctx.log.debug(
-        s"""Compiling:
-           |  javacOptions: ${javacOptions.map("'" + _ + "'").mkString(" ")}
-           |  scalacOptions: ${scalacOptions.map("'" + _ + "'").mkString(" ")}
-           |  sources: ${sources.map("'" + _ + "'").mkString(" ")}
-           |  classpath: ${compileClasspath.map("'" + _ + "'").mkString(" ")}
-           |  output: ${classesDir}"""
-          .stripMargin
-      )
-    }
-
-    reporter.foreach(_.start())
-
-    val consoleAppender = ConsoleAppender(
-      "ZincLogAppender",
-      ConsoleOut.printStreamOut(ctx.log.streams.err),
-      ctx.log.prompt.colored,
-      ctx.log.prompt.colored,
-      _ => None
+      javacOptions: Seq[String]
+  )(f: (ZincWorkerApi, JavaCompilerOptions) => A)(using
+      ctx: JvmWorkerApi.Ctx
+  ): A = {
+    val jOpts = JavaCompilerOptions(javacOptions)
+    val log = ctx.log
+    val zincCtx = ZincWorker.InvocationContext(
+      env = ctx.env,
+      dest = ctx.dest,
+      logDebugEnabled = log.debugEnabled,
+      logPromptColored = log.prompt.colored
     )
-    val loggerId = Thread.currentThread().getId.toString
-    val logger = SbtLoggerUtils.createLogger(loggerId, consoleAppender, zincLogLevel)
 
-    val maxErrors = reporter.map(_.maxErrors).getOrElse(CompileProblemReporter.defaultMaxErrors)
-
-    def mkNewReporter(mapper: (xsbti.Position => xsbti.Position) | Null) = reporter match {
-      case None =>
-        new ManagedLoggedReporter(maxErrors, logger) with RecordingReporter
-          with TransformingReporter(ctx.log.prompt.colored, mapper) {}
-      case Some(forwarder) =>
-        new ManagedLoggedReporter(maxErrors, logger)
-          with ForwardingReporter(forwarder)
-          with RecordingReporter
-          with TransformingReporter(ctx.log.prompt.colored, mapper) {}
-    }
-    val analysisMap0 = upstreamCompileOutput.map(c => c.classes.path -> c.analysisFile).toMap
+    def runWithLocalZinc() = {
+      val zincDeps =
+        ZincWorker.InvocationDependencies(
+          log = log,
+          consoleOut = ConsoleOut.printStreamOut(log.streams.err)
+        )
 
-    def analysisMap(f: VirtualFile): Optional[CompileAnalysis] = {
-      val analysisFile = f match {
-        case pathBased: PathBasedFile => analysisMap0.get(os.Path(pathBased.toPath))
-        case _ => None
-      }
-      analysisFile match {
-        case Some(zincPath) => fileAnalysisStore(zincPath).get().map(_.getAnalysis)
-        case None => Optional.empty[CompileAnalysis]
+      val api = new ZincWorkerApi {
```

Comment:
All the forwarding of `ZincWorkerApi` params to `zincLocalWorker` calls or `ZincWorkerRpcServer.ClientToServer` messages is pretty tedious. What if rather than making `ZincWorkerApi` a `trait`, we standardized on the `ZincWorkerRpcServer.ClientToServer` case classes as the way to pass data to both local and remote zinc workers? 

Then the entrypoint for `ZincWorker` can be a single method taking `ZincWorkerRpcServer.ClientToServer` and matching on the various cases, that can be re-used for both local and remote workers, without constantly having to convert between loose method parameters and bundled case class instances